core/iwasm/common/wasm_shared_memory.c

```diff
@@ -87,6 +92,43 @@ search_module(WASMModuleCommon *module)
     return NULL;
 }
 
+static void
+create_list_of_waiter_addresses(void *key, void *value, void *user_data)
+{
+    /* add waiter address to atomic wait list */
+    AtomicWaitAddressNode *wait_address_node;
+
+    if (!(wait_address_node =
+              wasm_runtime_malloc(sizeof(AtomicWaitAddressNode)))) {
+        return;
+    }
+    wait_address_node->addr = key;
+    bh_assert(bh_list_insert(user_data, wait_address_node) == BH_LIST_SUCCESS);
+}
+
+void
+notify_stale_threads_on_exception(WASMModuleInstanceCommon *module_inst)
+{
+    bh_list atomic_wait_address_list_head;
```

Comment:
does it have to be a linked list? I'd assume we know the exact size of the `wait_map`, so we can use array and only have a single allocation?